File: src/leetcode/check-if-number-has-equal-digit-count-and-digit-value/check-if-number-has-equal-digit-count-and-digit-value.js
/**
 * Leetcode #2283
 * Difficulty: Easy
 * https://leetcode.com/problems/check-if-number-has-equal-digit-count-and-digit-value/
 */
/**
 * @param {string} num
 * @return {boolean}
 */
export const digitCount = num => {
  const map = new Map();

  for (let i = 0; i < num.length; i += 1) {
    map.set(num[i], (map.get(num[i]) ?? 0) + 1);
  }

  for (let i = 0; i < num.length; i += 1) {
    const count = map.get(String(i)) ?? 0;

    if (count !== Number(num[i])) {
      return false;
    }
  }

  return true;
};

File: src/leetcode/check-if-number-has-equal-digit-count-and-digit-value/check-if-number-has-equal-digit-count-and-digit-value.rs
impl Solution {
    pub fn digit_count(num: String) -> bool {
        let mut map = std::collections::HashMap::new();

        for c in num.chars() {
            let digit: usize = c.to_digit(10).unwrap_or(0) as usize;
            *map.entry(digit).or_insert(0) += 1;
        }


        for (i, c) in num.chars().enumerate() {
            let count = c.to_digit(10).unwrap_or(0) as usize;
            let contains = map.contains_key(&i);

            if (contains && map[&i] != count) || (!contains && count != 0) {
                return false
            }
        }

        true
    }
}